apps/prairielearn/src/tests/groupExam.test.ts

```diff
@@ -0,0 +1,513 @@
+  assert.lengthOf(result.rows, 3);
+  return result.rows;
+}
+
+/**
+ * Switches active user and loads assessment. Asserts the number of CSRF tokens on the page
+ * after loading the given assessment URL.
+ */
+async function switchUserAndLoadAssessment(
+  studentUser: StudentUser,
+  assessmentUrl: string,
+  authUin: string,
+  numCsrfTokens: number,
+): Promise<{ $: cheerio.Root; csrfToken: string }> {
+  // Load config
+  config.authUid = studentUser.uid;
+  config.authName = studentUser.name;
+  config.authUin = authUin;
+
+  // Load assessment
+  const res = await fetch(assessmentUrl);
+  assert.isOk(res.ok);
+  const page = await res.text();
+  const $ = cheerio.load(page);
+
+  // Check for CSRF tokens
+  const elemList = $('form input[name="__csrf_token"]');
+  assert.lengthOf(elemList, numCsrfTokens);
+  assert.nestedProperty(elemList[0], 'attribs.value');
+  assert.isString(elemList.attr('value'));
+  const csrfToken = elemList.attr('value') as string; // guaranteed to be string by assertion
+
+  return { $, csrfToken };
+}
+
+/**
+ * Creates a new group in the given assessment as the user with the given CSRF token
+ */
+async function createGroup(
+  groupName: string,
+  csrfToken: string,
+  assessmentUrl: string,
+): Promise<cheerio.Root> {
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'create_group',
+      __csrf_token: csrfToken,
+      groupName: groupName,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+/**
+ * Joins a group in an assessment using the provided join code as the user with the given CSRF token
+ */
+async function joinGroup(
+  assessmentUrl: string,
+  joinCode: string,
+  csrfToken: string,
+): Promise<cheerio.Root> {
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      __action: 'join_group',
+      __csrf_token: csrfToken,
+      join_code: joinCode,
+    }),
+  });
+  assert.isOk(res.ok);
+  const $ = cheerio.load(await res.text());
+  return $;
+}
+
+describe('Group based exam assessments', function () {
+  this.timeout(20000);
+
+  before('set up testing server', helperServer.before(TEST_COURSE_PATH));
+  before('set authenticated user', function (callback) {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+    callback(null);
+  });
+
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function (callback) {
+    Object.assign(config, storedConfig);
+    callback(null);
+  });
+
+  describe('instructor access for exam assessment', function () {
+    step("should load the group tab for the first assessment's instructor URL", async function () {
+      // Get exam assessment URLs using ids from database
+      const result = await queryAsync(sql.select_group_work_exam_assessment, []);
```

Comment:
This and the following test appears to rely on the ordering of the assessments in the result set; while this seems to work, it's unnecessarily fragile. You could add an ORDER BY id` clause to the query, but that would still rely on them actually being inserted in the same order.

Rather than that, consider selecting the assessments by their human-readable IDs (e.g. `exam14-groupWork` / `exam15-groupWorkRoles`). This will be very resilent, and also much easier to use as a jumping off point to understand what the test does ("ah! this is an assessment identifier, I can go find this in teh `testCourse` directory.")